### PR TITLE
Add proper ASAN support to task switching

### PR DESCRIPTION
**Pull request by vchuravy**
When we are using ASAN we need to inform it when we switch stacks.
```
WARNING: ASan is ignoring requested __asan_handle_no_return: stack top: 0x7ffee525b000; bottom 0x7efe8a835000; size: 0x01005aa26000 (1101032218624)
False positive error reports may follow
For details see https://github.com/google/sanitizers/issues/189
```

Originally needed by https://github.com/JuliaLang/julia/pull/22631

@vtjnash any ideas for how to do this cleverly?

### CODE REVIEW

**Comment by vchuravy** [comment]
@vtjnash would you mind if I rewrite `jl_swap_fiber` and related function to operate on tasks instead of contexts?, so that instead of littering the ASAN support throughout the code base it is in one place? Similar to how Folly does this https://github.com/facebook/folly/commit/2ea64dd24946cbc9f3f4ac3f6c6b98a486c56e73

**Comment by vtjnash** [comment]
Sure, however you feel is convenient. I settled on this API to emphasize that it is expected to only access this field. But it’s not a hard restriction.

File: src/gc-debug.c [hunk 1]
```diff
@@ -579,27 +579,25 @@ static void gc_scrub_task(jl_task_t *ta)
...
+    else if (ta->stkbuf) {
+        low  = (char*)ta->stkbuf;
+        high = (char*)ta->stkbuf + ta->bufsz;
+    }
+    else
+        return;
+
     if (ptls == ptls2 && ta == ptls2->current_task) {
         // scan up to current `sp` for current thread and task
         char *low = (char*)jl_get_frame_addr();
```
Comment (vtjnash) [inline src/gc-debug.c]: in the future, we could probably store the precise answer during task switch too, and at the start of gc for most threads (that just leaves still-running threads in gc-safe regions)

File: src/task.c [hunk 2]
```diff
@@ -448,6 +462,7 @@ void JL_NORETURN throw_internal(jl_value_t *exception JL_MAYBE_UNROOTED)
         }
         assert(cur_block == eh->timing_stack);
 #endif
+        // todo sanitizer
```
Comment (vchuravy) [inline src/task.c]: @vtjnash in which stack does the exception handler run?

Comment (vtjnash) [reply src/task.c]: this is never a task switch, it's just a longjmp

**Comment by vtjnash** [comment]
Seems good

**Review by vtjnash — COMMENTED** [review]
Seems good

File: src/task.c (hunk 2 above)
Comment (vchuravy) [reply src/task.c]: We need to annotate that to which stack they are jumping otherwise ASAN doesn't like us.

**Comment by NHDaly** [comment]
Haha oops. Wellllll, it was fun anyway! 😂 And an interesting exercise!

**Comment by JeffBezanson** [comment]
Causes this compiler warning:
```
In file included from /home/jeff/src/julia/src/rtutils.c:23:0:
/home/jeff/src/julia/src/julia.h:95:20: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
 static inline void sanitizer_finish_switch_fiber() {}
                    ^
/home/jeff/src/julia/src/julia.h: In function ‘sanitizer_finish_switch_fiber’:
/home/jeff/src/julia/src/julia.h:95:20: warning: old-style function definition [-Wold-style-definition]
```

**Comment by vchuravy** [comment]
Sorry about that, noticed it after merging.

I have included a fix in #33259

**Pull request merged**